agent/eni/iphelperwrapper/iphelper_windows.go

```diff
@@ -0,0 +1,163 @@
+// We register a notification callback with Windows API for any changes in instance interfaces (NotifyIPInterfaceChange).
+// Initial notification received within timeout is considered as a confirmation of successful registration
+// notificationChannel is used for any communication between the callback and the consumer
+func (monitor *monitoringInterface) StartMonitor(notification chan int) error {
+	notificationChannel = notification
+
+	log.Info("Starting the monitor")
+	retVal, _, _ := funcNotifyIPInterfaceChange(syscall.AF_UNSPEC,
+		syscall.NewCallback(callback),
+		uintptr(unsafe.Pointer(&(monitor))),
+		1,
+		uintptr(unsafe.Pointer(&(monitor.notificationHandle))))
+
+	// If there is any error while registering callback then retVal will not be 0. Therefore, we will return with an error.
+	if retVal != 0 {
+		return errors.Errorf("error occured while calling Windows API : Error Msg : %s", syscall.Errno(retVal))
+	}
+
+	// Initial notification is sent immediately once the callback is registered.
+	// If the same is not received within a defined timeout then we assume an error and close the monitor
+initialNotificationLoop:
+	for {
+		select {
+		case event := <-notificationChannel:
+			if event == InitialNotification {
+				break initialNotificationLoop
+			}
+		case <-time.After(InitialNotificationTimeout):
+			log.Errorf("IPHelper : Initial notification not received within timeout. Closing the Monitor.")
+			if retErr := monitor.Close(); retErr != nil {
+				return errors.Wrapf(retErr, "initial notification not received as well")
+			}
+			return errors.Errorf("initial notification not received while starting monitor")
+		}
+	}
+
+	return nil
+}
+
+// This is the callback function invoked by NotifyIPInterfaceChange Windows API
+// The api returns the information about changed interface as a MIB_IPInterface_Row struct
+// The type of event can be inferred using the notificationType. Since we are concerned about addition of new ENIs,
+// therefore we will pass information to notification channel when the event is of interface addition i.e.1
+func callback(callerContext, row, notificationType uintptr) uintptr {
+
+	// Notification type 1 means addition of a device. InterfaceIndex would always be a positive value
+	if notificationType == 1 {
+		rowdata := *(*MibIPInterfaceRow)(unsafe.Pointer(row))
+		log.Debugf("Interface added with index : %v", rowdata.InterfaceIndex)
+		notificationChannel <- int(rowdata.InterfaceIndex)
+	} else if notificationType == 3 {
+		// Notification type = 3 means Initial notification. This block would be executed only once i.e. when agent starts
+		// Initial notification is sent immediately after registration.
+		// Only after sending this, any other types of notification can be sent
+		// Sending this notification on separate goroutine inorder to avoid deadlock with calling function
+		go func() {
+			time.Sleep(DelayInCallbackMessageForInitialNotification)
```

Comment:
Can this `Sleep` be eliminated ?

Reply:
It would be unadvisable to eliminate sleep from this section.
If the initial notification is triggered immediately after API registration then control cannot reach the corresponding select and application will assume an error.